File: FRONTEND/src/services/supplierService.ts
import type { Supplier, SupplierInput, SupplierListFilters, SupplierListResponse } from '@/models/supplier'
import { apiFetch } from '@/services/http'

export async function listSuppliers(filters: SupplierListFilters): Promise<SupplierListResponse> {
  const params = new URLSearchParams()
  params.set('page', String(filters.page))
  params.set('page_size', String(filters.page_size))

  if (filters.nome) params.set('nome', filters.nome)
  if (filters.cpf_cnpj) params.set('cpf_cnpj', filters.cpf_cnpj)
  if (typeof filters.ativo === 'boolean') params.set('ativo', String(filters.ativo))

  return apiFetch<SupplierListResponse>(`/fornecedores/?${params.toString()}`)
}

export async function getSupplierById(supplierId: number): Promise<Supplier> {
  return apiFetch<Supplier>(`/fornecedores/${supplierId}`)
}

export async function createSupplier(payload: SupplierInput): Promise<Supplier> {
  return apiFetch<Supplier>('/fornecedores/', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
}

export async function updateSupplier(supplierId: number, payload: SupplierInput): Promise<Supplier> {
  return apiFetch<Supplier>(`/fornecedores/${supplierId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
}

export async function deleteSupplier(supplierId: number): Promise<void> {
  return apiFetch<void>(`/fornecedores/${supplierId}`, { method: 'DELETE' })
}

File: API/app/services/supplier_service.py
from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.supplier import Fornecedor
from app.repositories.supplier_repository import SupplierRepository
from app.schemas.supplier import SupplierCreate, SupplierListResponse, SupplierRead, SupplierUpdate


class SupplierService:
    def __init__(self, session: AsyncSession) -> None:
        self.repository = SupplierRepository(session)

    async def list_suppliers(
        self,
        *,
        nome: str | None,
        cpf_cnpj: str | None,
        ativo: bool | None,
        page: int,
        page_size: int,
    ) -> SupplierListResponse:
        items, total = await self.repository.list_suppliers(
            nome=nome,
            cpf_cnpj=cpf_cnpj,
            ativo=ativo,
            page=page,
            page_size=page_size,
        )
        return SupplierListResponse(
            items=[SupplierRead.model_validate(item) for item in items],
            total=total,
            page=page,
            page_size=page_size,
        )

    async def get_supplier(self, supplier_id: int) -> SupplierRead | None:
        record = await self.repository.get_by_id(supplier_id)
        return None if record is None else SupplierRead.model_validate(record)

    async def create_supplier(self, payload: SupplierCreate) -> SupplierRead:
        record = Fornecedor(**payload.model_dump())
        return SupplierRead.model_validate(await self.repository.create(record))

    async def update_supplier(self, supplier_id: int, payload: SupplierUpdate) -> SupplierRead | None:
        record = await self.repository.get_by_id(supplier_id)
        if record is None:
            return None
        for field, value in payload.model_dump().items():
            setattr(record, field, value)
        return SupplierRead.model_validate(await self.repository.update(record))

    async def delete_supplier(self, supplier_id: int) -> bool:
        record = await self.repository.get_by_id(supplier_id)
        if record is None:
            return False
        try:
            await self.repository.delete(record)
        except Exception as exc:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Fornecedor vinculado a outros registros e nao pode ser excluido.",
            ) from exc
        return True